### ISSUE TITLE
Update Swagger Authentication

### ISSUE THREAD

**Issue by arnejduranovic** [CDCgov/prime-reportstream#10706]
### User Story

As the Swagger documentation page, I need to provide users a simple way to authenticate so the API calls will actually work on staging and production.

### Description/Use Case

Presently, the [API programmers guide](https://staging.reportstream.cdc.gov/resources/api/getting-started) covers TWO ways authentication can happen:
1. Method 1: Token-based authentication with a public/private key pair (Note: This method is the recommended best practice.)
2. Method 2: Shared secret API key (This is the Azure Function authentication scheme where you can generate a static API KEY in azure for a specific azure function and use that to authenticate. This is less secure than Method1 and is not encouraged)

The Swagger page has an "authorize" button. When the user clicks this, they should be able to at least use the second part of Method 1 to authenticate. Then, when they execute an endpoint in Swagger, the bearer token they entered will be sent in the header. Note, it would be nice to be able to do part one of Method 1 as well, requesting the token that is used in part 2 of Method 1, but they can separately use the swagger page for the `api/token` endpoint to do that part if needed. So, this means the user flow could potentially be:

1. User navigates to the `/api/token` swagger page and executes endpoint to get bearer token
2. User navigates to the swagger page containing the endpoint they want to use
3. User clicks "authorize" and enters the bearer token returned in step 1
4. User executes the desired endpoint and the bearer token is sent in the header

The downsides of this is that its kind of cumbersome on the user since the token will expire (every 5 min?). It would be really nice if the /api/token call was able to be configured in the same "authorize" window and can happen automatically. This seems to be how the default swagger Okta auth0 works?

### Risks/Impacts/Considerations

- Should be as user friendly as possible

This should also be tested locally. By default, the auth step will get skipped on a local env if no key is provided. If you provide the key in the header, the auth step WILL run. See flow below for more detail:

1. `AuthenticateClaims.authenticate` gets called at the top of the azure function
2. `getAccessToken` gets called at the top of `AuthenticateClaims.authenticate`
3. `isLocal` gets called right after `getAccessToken` and if the returned value of `getAccessToken` is not null auth will be executed. See `isLocal` for more details.

### Acceptance Criteria

- [X] User auth flow for Swagger designed and implemented (should be documented in swagger doc, separate story for this)
- [X] "Authorize" button works with server-server auth and tested locally (on the api keys page)
- [ ] Optionally, "Authorize" button works with x-functions key (on the api keys page)

Note: Authorization with azure API key will not show up on the swagger UI due to the auth method is deprecated by the API
 
Prerequisites:

hosting RS API docs swagger UI on RS server
#9302

**Comment by bishoyayoub** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Platform-6182b02547c1130010f459db/issues/CDCgov/prime-reportstream/10706?planning-poker) with Zenhub @arnejduranovic @jack-h-wang @JessicaWNava @JFU-NAVA-PBC @luis-pabon-tf @mkalish @thetaurean

**Comment by JFU-NAVA-PBC** [comment]
Quote some setup and okta provisioning steps here for convenience:

`This PR:

Enabled Authorization on API Docs swagger UI, the supported authorizations:

OAUTH2
Server2Server
Azure API Key

Test Steps:
1. Checkout the PR to local machine and start report stream server as per README
2. At the completion of the RS server start up, the swagger UI of API docs should be provisioned
3. Go to Azure Storage Explorer and copy the URL of swaggerUI index.html, on local setup, it is:
http://127.0.0.1:10000/devstoreaccount1/apidocs/index.html
4. Paste it into a browser, and you will see:
5. On top of the page: there is external Docs link - currently I put Report Stream API Programmer's Guide link there ( can also use the new UP documents "Swagger UI" mark down as the entry point )
6. Click "Authorize" button to bring up authorization panel where the following auth methods are available:
7. "primeSecurityServerToServer  (http, Bearer)" Method 1 - the process of obtaining the token using primeCLI commands or the python sample program remain as the recommended practice: just keep the legacy process since it worked well and mostly is only used by internal developers
8. "primeSecurityAPIKey (apiKey)" Method 2 - Only available on Azure
9. "primeSecurityOAUTH (OAuth2, authorizationCode)" - associated with OKTA Staging (hhs-prime.oktapreview.com)
10. For OAUTH2 to work, the swagger UI needs to be in a browser started with CORS disabled
e.g on Mac, start chrome as below:
`open -n -a /Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --args --user-data-dir="/tmp/chrome_dev_test" --disable-web-security`
Note, this PR mostly deal with swagger UI API Docs on a local setup, on PROD and LLEs, there is a ticket #10621 

On auth server side add the redirect url to the redirect URI list.

For local dev test:

`http://127.0.0.1:10000/devstoreaccount1/apidocs/oauth2-redirect.html`


## Changes
- *Added swagger annotations to enable authorizations supported: server 2 server, azure API key*


## Checklist

### Testing
- [X] Tested locally?
- [ ] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [ ] (For Changes to /frontend-react/...) Ran `npm run lint:write`? 
- [X] Added tests: manual tests: see the setup notes above, after Okta login close the swagger UI auth dialogue and call API  end points, e.g. GET public key for an org, to verify access token is checked, either turn on DEBUG logging and check the logging showing Oauth authentication passed, or put a break point at the authentication point in the end point:
e.g. AuthenticatedClaims.authenticate(request) <=== verify the header has "Authorization bearer: ....." with token value


`

**Comment by JFU-NAVA-PBC** [comment]
FYI,

For reviewers:

the Okta app I used for testing: hhs-prime.oktapreview.com, simple_report (localdev)

**Comment by JFU-NAVA-PBC** [comment]
How to test OAUTH2 authentication and authorization flow from swagger UI:

Steps:

1. OKTA side preparations:
2. Login into your OKTA account, and on your home dashboard, you will see apps who will participate in oauth2 authn & authz flow that are associated to you (your account), see screen shot:
3.[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBc0EyIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--41c0d9c822f36f12ef95d0951ad98d98d873b00a/image.png)

**Comment by JFU-NAVA-PBC** [comment]
NOTE:

If not doing OAUTH2 from a browser with CORS disabled, you will get 401 Error after OKTA login:

Auth ErrorError: response status is 401, error: invalid_client, description: Browser requests to the token endpoint must use Proof Key for Code Exchange.

See OKTA error message explanation link below:

https://support.okta.com/help/s/article/Browser-requests-to-the-token-endpoint-must-use-Proof-Key-for-Code-Exchange?language=en_US

Long story short: OKTA reject token exchange request from any non-server agent.
